File: kimdokyun/C++/boj_11054.cpp
#include<stdio.h>

int main() {
    int N,nums[1000],increase[1000]={},decrease[1000]={},result=0;
    scanf("%d",&N);
    for(int i=0;i<N;i++) scanf("%d",&nums[i]);
    for(int i=0;i<N;i++) {
        increase[i] = decrease[N-i-1] = 1;
        for(int j=0;j<i;j++) {
            increase[i] = nums[j] < nums[i] && increase[j]+1 > increase[i]?increase[j]+1:increase[i];
            decrease[N-i-1] = nums[N-j-1] < nums[N-i-1] && decrease[N-j-1]+1 > decrease[N-i-1]?decrease[N-j-1]+1:decrease[N-i-1];
        }
    }
    for(int i=0;i<N;i++) result = result < increase[i]+decrease[i]-1?increase[i]+decrease[i]-1:result;
    printf("%d",result);
}

File: kimdokyun/Python/boj_11054.py
N = int(input())
nums = list(map(int, input().split()))
increase = []
decrease = []
for i in range(N):
    increase.append(max([1]+[increase[j]+1 if nums[j] < nums[i] else 1 for j in range(i)]))
    decrease.insert(0, max([1]+[decrease[j]+1 if nums[j] < nums[-i-1] else 1 for j in range(-1, -i-1, -1)]))
print(max([increase[i]+decrease[i]-1 for i in range(N)]))